### ISSUE TITLE
MoE decode: an extra reader of a router weight (ffn_gate_inp) changes model output

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/llama.cpp#128]
# An extra reader of a router weight changes the model's output

Found while measuring the MoE look-ahead producer (PR #127, parked). The finding is
independent of look-ahead: **adding a `MUL_MAT` that reads the router weight
`ffn_gate_inp` changes the decode result**, while adding a neutral node of the same shape
does not. On this rig the change is large (perplexity 3.0613 -> 3.1377), deterministic,
and reproducible.

Any future graph-level feature that re-reads a router tensor will hit this.

## Environment

- Model: `Qwen3.8-Flash-Next-APEX-I-Mini` (`qwen4exp`, 48 MoE layers), 6 shards.
- RTX 5060 Ti, CUDA 13.2.1, this fork at `ec23a599f` plus the PR-P1 commit.
- `llama-perplexity -f ppl4.txt -b 1 -ub 1 -c 256 --chunks 1 --save-all-logits`,
  `--moe-expert-cache-size 84 -ngl 99 -fit off --n-cpu-moe 99
  --override-tensor per_layer_token_embd=CPU --load-mode none -t 6`.
- `-b 1 -ub 1` makes every ubatch a single decode row, so the producer is active.

## Isolation matrix

One build per row. `--moe-lookahead 8` in every row except the first.

| graph change | PPL | all-logits md5 |
| --- | --- | --- |
| none (baseline) | 3.0613 | `fc1da5b9e41d249e7da51791a934d561` |
| full producer (gate matmul + argsort + prefetch op) | 2.9434 | `8e0a9d10e9095ce7806f645e7edbba0e` |
| prediction only, prefetch op removed | 2.9434 | `8e0a9d10e9095ce7806f645e7edbba0e` |
| prediction reads a trunk tensor, no extra chain | 3.1849 | `4ba9f8e7a60ca9830b0dcbed073ffc3e` |
| gate `MUL_MAT` only, no argsort, no op | 3.1377 | `9bbf44eb3b98b5503ed2d1f9b028dda3` |
| neutral `COPY` node only | 3.0613 | `fc1da5b9e41d249e7da51791a934d561` |
| baseline with `GGML_CUDA_DISABLE_GRAPHS=1` | 3.0613 | `fc1da5b9e41d249e7da51791a934d561` |

All values are deterministic across reruns.

## What the matrix rules out

- **Not node count and not scheduling in general.** The neutral `COPY` node row is
  byte-identical to the baseline, so the backend is shape-neutral for a node that does not
  touch the router. The perturbation tracks readers of a router weight specifically.
- **Not the new op.** Removing `GGML_OP_MOE_PREFETCH` entirely leaves the diverging md5
  unchanged, so the op contributes nothing to the number.
- **Not lost CUDA graph capture.** Disabling graphs on the unmodified baseline reproduces
  the baseline exactly.
- **Not the extra activation chain.** Using an already-live trunk tensor (no extra
  `build_hc_mix` chain) still diverges.

## Hypothesis

The decode-path decision is count-based. The MoE grouped-decode certificate proves the
graph from backend `use_counts`, and it already had to be taught to ignore the new op's
sources - without that exclusion it rejects the graph outright:

```
E moe-cache: grouped decode certificate failed: semantic_group=1 layout=1 domain=1
             tensor=blk.1.ffn_gate_exps.weight graph=unproven(14)
E ggml_backend_sched_graph_compute_async failed with error -1
```

That is the same machinery named in the census path
(`moe_candidate_discover_route` / `moe_candidate_active_reverse` / `use_counts`). An extra
reader of `ffn_gate_inp` changes the counts for a router tensor, which plausibly
reclassifies the layer's execution path and therefore the arithmetic.

Unconfirmed. The next step would be one probe of the census verdict per layer with and
without the extra reader, before attempting any fix.

## Suggested fix location

An exclusion mechanism for prediction-subgraph nodes in the route/candidate census, next
to the existing `GGML_OP_MOE_PREFETCH` use-count exclusion in `ggml.c`. This is not a
prediction-math problem, and it is worth fixing if the fork keeps growing graph-level
features that read router tensors.

## Provenance

Measured during PR #127 (`feat/moe-lookahead-p1`, commit `4f681738b`). The full write-up
including the throughput and pool-authority blockers is in
`docs/moe-lookahead-design.md`, section "PR-P1 outcome (measured)", commit `495da1938`.

**Comment by ddvnguyen** [comment]
## Cleared on this configuration - with a working instrument this time

The claim: an extra reader of the router weight (`ffn_gate_inp`) changes model output. The token-identity
check previously cited in support of the opposite conclusion was **vacuous** (it compared empty
strings) and has been retracted. Here is a check that actually reads what the model produced.

**Why it is not vacuous.** `build_moe_lookahead` returns early when `cparams.moe_lookahead <= 0`
(`src/llama-graph.cpp:2095`), so `--moe-lookahead 0` removes the extra reader rather than setting its
width to zero. The ON/OFF comparison is therefore a real A/B on the reader's presence.

**Test 1 - greedy token identity.** POST `/completion`, `stream: false`, `temperature: 0`, `top_k: 1`,
fixed `--seed`, reading `.content` from the response body (the streaming channel is what broke the old
check). Greedy is maximally sensitive: any perturbation that flips an argmax anywhere diverges.

| arm | look-ahead | chars | output sha256 |
|---|---|---|---|
| la0 | off | 648 | `73253b673b79b7ab` |
| la0b | off (repeat) | 648 | `73253b673b79b7ab` |
| la8 | on (width 8) | 648 | `73253b673b79b7ab` |

The repeat arm is a determinism control across two separate server launches; it passes, so the
instrument is trustworthy, and the ON arm is byte-identical to it.

**Test 2 - perplexity**, because greedy only catches an argmax flip and would miss a pure
distribution shift. `llama-perplexity`, 15 chunks x 512 tokens of real prose, same corpus and seed:

```
look-ahead off       PPL = 14.7350 +/- 0.71885
look-ahead off (x2)  PPL = 14.7350 +/- 0.71885
look-ahead on        PPL = 14.7350 +/- 0.71885
```

Per-chunk values match to four decimals in all three arms (`[1]8.6064 ... [15]14.7350`), i.e.
logits-level identity rather than "no divergence observed".

**Scope.** One prompt for the token test, one corpus for the PPL test, greedy temperature 0. This
clears the blocker on these paths. It does not prove the reader is free under sampling or on other
architectures.

**But the ON arm does not run at the recommended cache size at all** - see #129: at
`--moe-expert-cache-size 53` the look-ahead aborts the process. That is now the live blocker on this
feature, and it is a different defect from this one.
